File: integration_test_app/ios/Runner/TestRtcEnginePlugin.m
#import <Foundation/Foundation.h>
#import "TestRtcEnginePlugin.h"

@interface TestRtcEnginePlugin : NSObject <RtcEnginePlugin>
@property(nonatomic) bool isRtcEngineCreated;
@property(nonatomic) bool isRtcEngineDestroy;
@property(nonatomic) AgoraRtcEngineKit *_Nullable rtcEngine;

@end

@implementation TestRtcEnginePlugin

- (void)onRtcEngineCreated:(AgoraRtcEngineKit * _Nullable)rtcEngine {
    self.isRtcEngineCreated = true;
    self.isRtcEngineDestroy = false;
    self.rtcEngine = rtcEngine;
}

- (void)onRtcEngineDestroyed {
    self.isRtcEngineCreated = false;
    self.isRtcEngineDestroy = true;
}

@end

@interface TestRtcEnginePluginMethodChannel()
@property(nonatomic) TestRtcEnginePlugin *currentRtcEnginePlugin;
@end

@implementation TestRtcEnginePluginMethodChannel

- (void)setUp:(NSObject<FlutterBinaryMessenger>*)messenger {
    FlutterMethodChannel* methodChannel = [FlutterMethodChannel
                                            methodChannelWithName:@"agora_rtc_engine/integration_test/rtc_engine_plugin"
                                            binaryMessenger:messenger];

    [methodChannel setMethodCallHandler:^(FlutterMethodCall* call, FlutterResult result) {
        if ([@"registerRtcEnginePlugin" isEqualToString:call.method]) {
            self.currentRtcEnginePlugin = [TestRtcEnginePlugin new];
            [RtcEnginePluginRegistrant register:self.currentRtcEnginePlugin];
            result(@true);
        } else if ([@"unregisterRtcEnginePlugin" isEqualToString:call.method]) {
            [RtcEnginePluginRegistrant unregister:self.currentRtcEnginePlugin];
            result(@true);
        } else if ([@"isRtcEngineCreated" isEqualToString:call.method]) {
            result(@([[self currentRtcEnginePlugin] isRtcEngineCreated]));
        } else if ([@"isRtcEngineDestroyed" isEqualToString:call.method]) {
            result(@([[self currentRtcEnginePlugin] isRtcEngineDestroy]));
        } else if ([@"getRtcEngineNativeHandleFromPlugin" isEqualToString:call.method]) {
            result(@((intptr_t)[[[self currentRtcEnginePlugin] rtcEngine] getNativeHandle]));
        } else {
            result(FlutterMethodNotImplemented);
          }
    }];
}

@end

File: integration_test_app/android/app/src/main/kotlin/io/agora/integration_test_app/integration_test_app/TestRtcEnginePlugin.kt
package io.agora.integration_test_app.integration_test_app

import io.agora.rtc.RtcEngine
import io.agora.rtc.base.RtcEnginePlugin
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

class TestRtcEnginePluginMethodChannel : MethodChannel.MethodCallHandler {

  private var currentRtcEnginePlugin: TestRtcEnginePlugin? = null

  fun setUp(binaryMessenger: BinaryMessenger) {
    MethodChannel(binaryMessenger, "agora_rtc_engine/integration_test/rtc_engine_plugin").apply {
      setMethodCallHandler(this@TestRtcEnginePluginMethodChannel)
    }
  }

  override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
    when (call.method) {
      "registerRtcEnginePlugin" -> {
        currentRtcEnginePlugin = TestRtcEnginePlugin()
        RtcEnginePlugin.register(currentRtcEnginePlugin!!)
        result.success(true)
      }
      "unregisterRtcEnginePlugin" -> {
        RtcEnginePlugin.unregister(currentRtcEnginePlugin!!)
        result.success(true)
      }
      "isRtcEngineCreated" -> {
        result.success(currentRtcEnginePlugin?.isRtcEngineCreated)
      }
      "isRtcEngineDestroyed" -> {
        result.success(currentRtcEnginePlugin?.isRtcEngineDestroyed)
      }
      "getRtcEngineNativeHandleFromPlugin" -> {
        result.success(currentRtcEnginePlugin?.rtcEngine?.nativeHandle)
      }
      else -> {
        result.notImplemented()
      }
    }
  }

}

class TestRtcEnginePlugin : RtcEnginePlugin {

  var isRtcEngineCreated: Boolean = false
  var isRtcEngineDestroyed: Boolean = false
  var rtcEngine: RtcEngine? = null

  override fun onRtcEngineCreated(rtcEngine: RtcEngine?) {
    isRtcEngineCreated = true
    isRtcEngineDestroyed = false
    this.rtcEngine = rtcEngine
  }

  override fun onRtcEngineDestroyed() {
    isRtcEngineCreated = false
    isRtcEngineDestroyed = true
  }
}
